flair/embeddings/base.py

```diff
@@ -149,3 +165,616 @@ def forward(self, tensors: List[torch.Tensor]) -> torch.Tensor:
+                subword_start_idx = subword_end_idx
+                if self.subtoken_pooling == "first":
+                    final_embedding = current_embeddings[:, 0]
+                elif self.subtoken_pooling == "last":
+                    final_embedding = current_embeddings[:, -1]
+                elif self.subtoken_pooling == "first_last":
+                    final_embedding = torch.cat([current_embeddings[:, 0], current_embeddings[:, -1]], dim=1)
+                elif self.subtoken_pooling == "mean":
+                    final_embedding = current_embeddings.mean(dim=1)
+                else:
+                    raise ValueError(f"subtoken pooling method: `{self.subtoken_pooling}` is not implemented")
+
+                if self.layer_mean:
+                    final_embedding = final_embedding.mean(dim=0)
+                else:
+                    final_embedding = torch.flatten(final_embedding)
+
+                token.set_embedding(self.name, final_embedding)
+
+    def _add_embeddings_to_sentences(self, sentences: List[Sentence]):
+        tokenized_sentences, all_token_subtoken_lengths = self._gather_tokenized_strings(sentences)
+
+        # encode inputs
+        batch_encoding = self.tokenizer(
+            tokenized_sentences,
+            stride=self.stride,
+            return_overflowing_tokens=self.allow_long_sentences,
+            truncation=self.truncate,
+            padding=True,
+            return_tensors="pt",
+        )
+
+        input_ids, model_kwargs = self._build_transformer_model_inputs(batch_encoding, tokenized_sentences, sentences)
+
+        gradient_context = torch.enable_grad() if (self.fine_tune and self.training) else torch.no_grad()
+
+        with gradient_context:
+            hidden_states = self.model(input_ids, **model_kwargs)[-1]
+
+            # make the tuple a tensor; makes working with it easier.
+            hidden_states = torch.stack(hidden_states)
+
+            # only use layers that will be outputted
+            hidden_states = hidden_states[self.layer_indexes, :, :]
+
+            if self._try_document_embedding_shortcut(hidden_states, sentences):
+                return
+
+            if self.allow_long_sentences:
+                sentence_hidden_states = self._combine_strided_sentences(
+                    hidden_states,
+                    sentence_parts_lengths=torch.unique(
+                        batch_encoding["overflow_to_sample_mapping"],
+                        return_counts=True,
+                        sorted=True,
+                    )[1].tolist(),
+                )
+            else:
+                sentence_hidden_states = list(hidden_states.permute((1, 0, 2, 3)))
+
+            # remove padding tokens
+            sentence_hidden_states = [
+                sentence_hidden_state[:, : len(subtokens) + 1, :]
+                for (subtokens, sentence_hidden_state) in zip(tokenized_sentences, sentence_hidden_states)
+            ]
```

Comment:
This causes problems if a sentence has more subtokens than tokens. Padding tokens should be removed to subtoken length:

```suggestion
            sentence_hidden_states = [
                sentence_hidden_state[:, : sum(token_subtoken_lengths) + 1, :]
                for (token_subtoken_lengths, sentence_hidden_state) in zip(all_token_subtoken_lengths, sentence_hidden_states)
            ]
```

Reply:
Very good catch, thank you.
My fix is slightly different: `all_token_subtoken_lengths` is only filled if it is a token embedding, as document embeddings don't need the subtoken_token mapping and shouldn't create that overhead.
besides that, I noticed that if the last subtoken doesn't map to a token it will be discarded, so I rather store the full subtoken length and take that one. Hence we can ensure that TransformerDocumentEmbeddings with pad_last always have the right token as last.